File: server/src/macros.ts
//! User‑macro extraction and expansion.
//!
//! LaTeX documents can define custom commands via several mechanisms:
//!
//! ```tex
//! \newcommand{\R}{\mathbb{R}}           % zero‑argument
//! \newcommand*{\R}{\mathbb{R}}          % starred (no paragraph breaks in body)
//! \renewcommand{\R}{\mathbb{R}}         % redefine existing
//! \providecommand{\R}{\mathbb{R}}       % define only if undefined
//! \def\R{\mathbb{R}}                    % plain‑TeX definition
//! \DeclareMathOperator{\div}{div}       % math‑operator wrapper
//! ```
//!
//! ## Usage
//!
//! ```ts
//! const macros = extractMacros(documentText);
//! const expanded = expand("$\\R^2$", macros);  // → $\\mathbb{R}^2$
//! ```
//!
//! ## Whitelist
//!
//! Only the six defining commands above are recognised.  Arbitrary LaTeX
//! patterns that *look* like a definition (e.g. `\end{equation}\eps{3}`)
//! are NOT captured — the scanner must match one of the defining‑command
//! names or the match is discarded.
//!
//! ## Brace handling
//!
//! Macro bodies and arguments may contain nested braces (e.g.
//! `\newcommand{\x}{\sqrt{\frac{a}{b}}}`).  Regex‑based parsing cannot
//! handle arbitrary nesting, so this module walks the source with an
//! explicit brace counter for bodies and arguments.

export type MacroMap = Record<string, string>;

// ── defining‑command whitelist ─────────────────────────────────────────
// Order matters: the longer / starred forms must be tried before their
// bare prefix (`newcommand*` before `newcommand`).

const DEFINING_CMDS = [
  "newcommand*", "newcommand",
  "renewcommand*", "renewcommand",
  "providecommand*", "providecommand",
  "def",
  "DeclareMathOperator",
];

// ── brace-balanced helpers ────────────────────────────────────────────

/** Read a balanced `{…}` starting at index `at` (which must point at `{`).
 *  Returns the inner content (without the outer braces) and the offset
 *  just past the closing `}`.  Returns `null` if the braces never balance. */
function readBalancedBraces(src: string, at: number): { body: string; end: number } | null {
  if (src[at] !== "{") return null;
  let depth = 1;
  for (let i = at + 1; i < src.length; i++) {
    const ch = src[i];
    if (ch === "\\") { i++; continue; }            // skip escaped char
    if (ch === "{") depth++;
    else if (ch === "}") { depth--; if (depth === 0) return { body: src.slice(at + 1, i), end: i + 1 }; }
  }
  return null;
}

/** Skip spaces and tabs starting at `at`.  Returns the next non‑space
 *  offset. */
function skipWs(src: string, at: number): number {
  let i = at;
  while (i < src.length && (src[i] === " " || src[i] === "\t")) i++;
  return i;
}

// ── extraction ────────────────────────────────────────────────────────

/** Try to parse a macro definition starting at `at` (which must point at
 *  the leading `\`).  Returns `{ cmd, name, arity, bodyOpen }` on success
 *  where `bodyOpen` is the offset of the body's opening `{`, or `null` if
 *  the text at `at` is not one of the defining commands. */
function readMacroDef(src: string, at: number):
  { cmd: string; name: string; arity: number; bodyOpen: number } | null {
  // The character at `at` must be `\`.
  if (src[at] !== "\\") return null;

  // Try each defining command in turn.  The longest match wins so starred
  // forms (`\newcommand*`) are not misread as their bare prefix.
  let cmd: string | null = null;
  let after: number = at + 1;
  for (const c of DEFINING_CMDS) {
    if (src.startsWith(c, after) && !/[A-Za-z]/.test(src[after + c.length] ?? "")) {
      cmd = c;
      after += c.length;
      break;
    }
  }
  if (cmd === null) return null;

  after = skipWs(src, after);

  // Read the macro name: either `{\name}` (newcommand family) or `\name`
  // (plain \def).  Both forms are recognised for every command.
  let name: string;
  if (src[after] === "{") {
    const close = src.indexOf("}", after + 1);
    if (close < 0 || src[after + 1] !== "\\") return null;
    name = src.slice(after + 2, close);
    after = close + 1;
  } else if (src[after] === "\\") {
    let end = after + 1;
    while (end < src.length && /[A-Za-z@]/.test(src[end])) end++;
    name = src.slice(after + 1, end);
    after = end;
  } else {
    return null;
  }

  // Optional `[N]` arity.
  after = skipWs(src, after);
  let arity = 0;
  if (src[after] === "[") {
    const close = src.indexOf("]", after + 1);
    if (close < 0) return null;
    arity = Number(src.slice(after + 1, close));
    after = close + 1;
  }

  // Body must follow as `{…}`.
  after = skipWs(src, after);
  if (src[after] !== "{") return null;
  return { cmd, name, arity, bodyOpen: after };
}

/** Extract all user‑defined macros from a LaTeX document text. */
export function extractMacros(text: string): MacroMap {
  const out: MacroMap = {};
  let i = 0;
  while (i < text.length) {
    if (text[i] === "\\") {
      const def = readMacroDef(text, i);
      if (def) {
        const body = readBalancedBraces(text, def.bodyOpen);
        if (body) {
          let bodyStr = body.body;
          if (def.cmd === "DeclareMathOperator") {
            bodyStr = `\\operatorname{${bodyStr}}`;
          }
          out[def.name] = bodyStr;
          out[`__arity__${def.name}`] = String(def.arity);
          i = body.end;
          continue;
        }
      }
    }
    i++;
  }
  return out;
}

/**
 * Merge two macro maps, returning a new map.
 * `overrides` take precedence — macros defined there overwrite same‑named
 * macros in `base`.
 */
export function mergeMacros(base: MacroMap, overrides: MacroMap): MacroMap {
  return { ...base, ...overrides };
}

// ── expansion ─────────────────────────────────────────────────────────

/** Read one macro argument starting at `at`: either a brace-delimited
 *  `{…}` (with arbitrary nesting) or a single non-whitespace token.
 *  Returns `{ arg, end }` on success. */
function readArg(src: string, at: number): { arg: string; end: number } | null {
  let i = skipWs(src, at);
  if (i >= src.length) return null;
  if (src[i] === "{") {
    const r = readBalancedBraces(src, i);
    if (!r) return null;
    return { arg: r.body, end: r.end };
  }
  // Bare token: read until whitespace, brace, or end of input.
  let end = i;
  while (end < src.length && src[end] !== " " && src[end] !== "\t" && src[end] !== "{") end++;
  return { arg: src.slice(i, end), end };
}

/**
 * Expand every known macro in `source`, returning the expanded LaTeX string.
 *
 * Macros without arguments are substituted via `\name\b` boundary‑match.
 * Macros with arguments honour `#1`, `#2`, … placeholders; each argument may
 * be brace‑delimited (`{arg}`, with arbitrary nesting) or a single
 * non‑whitespace token (`\S`).
 */
export function expand(source: string, macros: MacroMap): string {
  let out = source;
  for (const [name, raw] of Object.entries(macros)) {
    if (name.startsWith("__arity__")) continue;
    const arity = Number(macros[`__arity__${name}`] ?? "0");
    const head = `\\${name}`;

    // Walk the source left-to-right, replacing each occurrence.  Using a
    // manual scan rather than a regex lets us call readArg (which
    // understands balanced braces) for each match.
    let result = "";
    let cursor = 0;
    while (cursor < out.length) {
      const idx = out.indexOf(head, cursor);
      if (idx < 0) {
        result += out.slice(cursor);
        break;
      }
      // Word boundary: next char must not be an identifier continuation.
      const nextCh = out[idx + head.length];
      if (nextCh && /[A-Za-z@]/.test(nextCh)) {
        result += out.slice(cursor, idx + head.length);
        cursor = idx + head.length;
        continue;
      }
      result += out.slice(cursor, idx);

      if (arity === 0) {
        result += raw;
        cursor = idx + head.length;
        continue;
      }

      // N-argument macro: consume `arity` arguments after the head.
      let pos = idx + head.length;
      const args: string[] = [];
      let ok = true;
      for (let a = 0; a < arity; a++) {
        const r = readArg(out, pos);
        if (!r) { ok = false; break; }
        args.push(r.arg);
        pos = r.end;
      }
      if (!ok) {
        // Could not parse enough arguments — leave the call site untouched.
        result += out.slice(idx, pos);
        cursor = pos;
        continue;
      }
      let expansion = raw;
      for (let a = 1; a <= arity; a++) {
        expansion = expansion.replaceAll(`#${a}`, args[a - 1]);
      }
      result += expansion;
      cursor = pos;
    }
    out = result;
  }
  return out;
}

File: latex-index/src/macros.rs
//! Macro extraction — port of `server/src/macros.ts`.
//!
//! Recognises the same six defining commands and uses an explicit
//! brace-counter so macro bodies with arbitrary nesting (e.g.
//! `\newcommand{\x}{\sqrt{\frac{a}{b}}}`) parse correctly.  See
//! `docs/plan-ref-cite-hover.md` Section 5.3.

use crate::index::{Index, MacroDef};
use std::path::Path;

/// Order matters: the longer / starred forms must come before their bare
/// prefix (`newcommand*` before `newcommand`).
const DEFINING_CMDS: &[&str] = &[
    "newcommand*",
    "newcommand",
    "renewcommand*",
    "renewcommand",
    "providecommand*",
    "providecommand",
    "def",
    "DeclareMathOperator",
];

// ── brace-balanced helpers ─────────────────────────────────────────────

/// Read a balanced `{…}` starting at `at` (which must point at `{`).
/// Returns `(inner_start, inner_end, end)` or `None`.
fn read_balanced(text: &[u8], at: usize) -> Option<(usize, usize, usize)> {
    if text.get(at) != Some(&b'{') {
        return None;
    }
    let mut depth = 1i32;
    let mut i = at + 1;
    while i < text.len() {
        match text[i] {
            b'\\' => {
                i = (i + 2).min(text.len());
                continue;
            }
            b'{' => {
                depth += 1;
                i += 1;
            }
            b'}' => {
                depth -= 1;
                i += 1;
                if depth == 0 {
                    return Some((at + 1, i - 1, i));
                }
            }
            _ => i += 1,
        }
    }
    None
}

fn skip_ws(text: &[u8], at: usize) -> usize {
    let mut i = at;
    while i < text.len() && (text[i] == b' ' || text[i] == b'\t') {
        i += 1;
    }
    i
}

fn starts_with_word(text: &[u8], at: usize, word: &[u8]) -> bool {
    at + word.len() <= text.len() && &text[at..at + word.len()] == word
}

fn next_char_ident(b: u8) -> bool {
    b.is_ascii_alphabetic() || b == b'@'
}

// ── one macro definition ───────────────────────────────────────────────

#[derive(Debug, Clone, Copy)]
struct MacroSite {
    cmd: &'static str,
    name: &'static str,
    arity: u32,
    body_open: usize,
}

/// Try to parse a macro definition starting at `at` (the leading `\`).
/// Returns the macro descriptor or `None`.
fn read_macro_def(text: &[u8], at: usize) -> Option<MacroSite> {
    if text.get(at) != Some(&b'\\') {
        return None;
    }
    let mut cmd: Option<&'static str> = None;
    let mut after = at + 1;
    for &c in DEFINING_CMDS {
        let cb = c.as_bytes();
        if starts_with_word(text, after, cb)
            && !next_char_ident(*text.get(after + cb.len()).unwrap_or(&0))
        {
            cmd = Some(c);
            after += cb.len();
            break;
        }
    }
    let cmd = cmd?;
    after = skip_ws(text, after);

    // Macro name: either `{\name}` (newcommand family) or `\name` (plain \def).
    // Both forms are accepted for every command.
    let name_start: usize;
    let name_end: usize;
    if text.get(after) == Some(&b'{') {
        let close_idx = find_char(text, after + 1, b'}')?;
        if text.get(after + 1) != Some(&b'\\') {
            return None;
        }
        name_start = after + 2;
        name_end = close_idx;
        after = close_idx + 1;
    } else if text.get(after) == Some(&b'\\') {
        let mut end = after + 1;
        while end < text.len() && (text[end].is_ascii_alphabetic() || text[end] == b'@') {
            end += 1;
        }
        name_start = after + 1;
        name_end = end;
        after = end;
    } else {
        return None;
    }

    let name = std::str::from_utf8(&text[name_start..name_end]).ok()?;
    if name.is_empty() {
        return None;
    }

    // Optional `[N]` arity.
    after = skip_ws(text, after);
    let mut arity: u32 = 0;
    if text.get(after) == Some(&b'[') {
        let close = find_char(text, after + 1, b']')?;
        let n: u32 = std::str::from_utf8(&text[after + 1..close])
            .ok()?
            .trim()
            .parse()
            .ok()?;
        arity = n;
        after = close + 1;
    }

    // Body must follow as `{…}`.
    after = skip_ws(text, after);
    if text.get(after) != Some(&b'{') {
        return None;
    }

    // SAFETY: we copy `name` into a `&'static str` via leaking the bytes for
    // the duration of this call only.  Easier: just return `(name_owned,
    // cmd, arity, body_open)` by leaking the slice.  In practice we never
    // hold on to these across calls, so a `String` clone is fine.
    Some(MacroSite {
        cmd,
        // SAFETY: `name` lives until end of function; the lifetime here is
        // fine because MacroSite is consumed immediately in `extract`.
        name: Box::leak(name.to_string().into_boxed_str()),
        arity,
        body_open: after,
    })
}

fn find_char(text: &[u8], from: usize, target: u8) -> Option<usize> {
    let mut i = from;
    while i < text.len() {
        if text[i] == target {
            return Some(i);
        }
        i += 1;
    }
    None
}

// ── extraction ─────────────────────────────────────────────────────────

/// Parse `text` (contents of a `.tex` file at `path`) and insert every
/// `\newcommand`/`\def`/… definition into `index`.  Replaces any prior
/// entries from `path`.
pub fn extract_macros(text: &str, path: &Path, index: &Index) {
    let bytes = text.as_bytes();
    index.macros.retain(|_, v| v.file != path);

    let mut i = 0usize;
    while i < bytes.len() {
        if bytes[i] == b'\\' {
            if let Some(site) = read_macro_def(bytes, i) {
                if let Some((a, b, end)) = read_balanced(bytes, site.body_open) {
                    let mut body_str = std::str::from_utf8(&bytes[a..b])
                        .unwrap_or("")
                        .to_string();
                    if site.cmd == "DeclareMathOperator" {
                        body_str = format!("\\operatorname{{{}}}", body_str);
                    }
                    let def = MacroDef {
                        name: site.name.to_string(),
                        file: path.to_path_buf(),
                        body: body_str,
                        arity: site.arity,
                    };
                    index.macros.insert(def.name.clone(), def);
                    i = end;
                    continue;
                }
                // Drop the leaked `name` slice by re-using it as static.
                // (It's a small allocation per failed parse; acceptable.)
            }
        }
        i += 1;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn newcommand_basic() {
        let idx = Index::new();
        let text = r#"\newcommand{\R}{\mathbb{R}}"#;
        extract_macros(text, Path::new("a.tex"), &idx);
        let m = idx.macros.get("R").unwrap();
        assert_eq!(m.body, "\\mathbb{R}");
        assert_eq!(m.arity, 0);
    }

    #[test]
    fn newcommand_starred() {
        let idx = Index::new();
        let text = r#"\newcommand*{\H}{\mathbb{H}}"#;
        extract_macros(text, Path::new("a.tex"), &idx);
        assert!(idx.macros.get("H").is_some());
    }

    #[test]
    fn newcommand_with_arity() {
        let idx = Index::new();
        let text = r#"\newcommand{\norm}[1]{\left\|#1\right\|}"#;
        extract_macros(text, Path::new("a.tex"), &idx);
        let m = idx.macros.get("norm").unwrap();
        assert_eq!(m.arity, 1);
        assert!(m.body.contains("#1"));
    }

    #[test]
    fn def_with_bare_name() {
        let idx = Index::new();
        let text = r#"\def\R{\mathbb{R}}"#;
        extract_macros(text, Path::new("a.tex"), &idx);
        let m = idx.macros.get("R").unwrap();
        assert_eq!(m.body, "\\mathbb{R}");
    }

    #[test]
    fn declare_math_operator() {
        let idx = Index::new();
        let text = r#"\DeclareMathOperator{\div}{div}"#;
        extract_macros(text, Path::new("a.tex"), &idx);
        let m = idx.macros.get("div").unwrap();
        assert_eq!(m.body, "\\operatorname{div}");
    }

    #[test]
    fn nested_braces_in_body() {
        let idx = Index::new();
        let text = r#"\newcommand{\x}{\sqrt{\frac{a}{b}}}"#;
        extract_macros(text, Path::new("a.tex"), &idx);
        let m = idx.macros.get("x").unwrap();
        assert_eq!(m.body, "\\sqrt{\\frac{a}{b}}");
    }
}
